### DISCUSSION TITLE
How to make a combobox a reusable component in vue?

### DISCUSSION THREAD

**Discussion by StefanGhiban** [tailwindlabs/headlessui#1328] [Help]
Hi! I was trying to create a combobox as a reusable component but I can't figure out how I should do it. As the combobox itself doesnt have an @change function, how should I be able to make it as a reusable component. How to make a combobox update the state of a parent element? Thak you.

**Comment by Eliophot** [comment]
Hello, you can use `@update:modelValue` in combobox for that, I think.

**Reply by StefanGhiban** [reply]
Yes, I have seen in a thread, but then, what do you pass up? The value itself is on the combobox option, not the main combo component. $event.target.value doesnt seem to work.

**Comment by Eliophot** [comment]
https://vuejs.org/guide/components/events.html#usage-with-v-model
I will try to make one...

**Reply by StefanGhiban** [reply]
Yes, I understand the documentation and I have no problems making other types of reusable components, such as inputs, checkboxes etc. But this combobox component has a v-model of its own and I do not understand how should I apply the documentation from vue to this component. I have tried using emits and model value.

**Reply by Eliophot** [reply]
Indeed, I just tested, I had to do this for it to work : 
```
    <Combobox
        as="div"
        v-model="selectedItem"
        @update:modelValue="$emit('update:modelValue', selectedItem.id)"
    >
```

**Reply by StefanGhiban** [reply]
Yes, made it work similarly, but not having :value="modelValue" feels extremely weird tbh. Thanks for the help.

**Reply by StefanGhiban** [reply]
Still cant make it to work properly tho, what is the v-model here, how do you pass an initial value into the select?

**Reply by Eliophot** [reply]
```
<CommonFormSelect :data="companies" v-model="company.id" />
```

**Reply by StefanGhiban** [reply]
Your example doesnt work, people.id would fail because people would be an array. And if the v-model is done like this, how would you change the v-model from the combobox to take into account this v-model? If you have a full working repo for this reusable combox box, please send it to me, as these small pieces do not make any sense to me, sorry.

**Reply by Eliophot** [reply]
in the component : 
```
    props: {
        modelValue: {
            type: String,
            default: null,
        },
        data: {
            type: Object,
            default: null,
        },
    },
    emits: ["update:modelValue"],
```
in setup's component
```
    const selectedItem = ref(props.data[0]);

    props.data.forEach((item) => {
        if (item.id === props.modelValue) {
            selectedItem.value = item;
        }
    });

    onMounted(() => {
       getCurrentInstance().emit("update:modelValue", selectedItem.value.id);
    });
```
I think we can do better ^^

**Reply by Eliophot** [reply]
> Your example doesnt work, people.id would fail because people would be an array. And if the v-model is done like this, how would you change the v-model from the combobox to take into account this v-model? If you have a full working repo for this reusable combox box, please send it to me, as these small pieces do not make any sense to me, sorry.

yes sorry to fast response ^^

**Reply by StefanGhiban** [reply]
Yea, it is a bit convoluted. If we would compare this approach to the one from the vue docs, this is harder to read. Clearly there is an issue in how these components are constructed as making them reusable is very hard in my opinion.

**Reply by StefanGhiban** [reply]
> > Your example doesnt work, people.id would fail because people would be an array. And if the v-model is done like this, how would you change the v-model from the combobox to take into account this v-model? If you have a full working repo for this reusable combox box, please send it to me, as these small pieces do not make any sense to me, sorry.
> 
> yes sorry to fast response ^^

no worries

**Comment by adamleemiller** [comment]
I know this is long overdue here but I am trying to write a custom component myself using Vue and finding this extremely difficult. I have gotten to the point where I can set the selected option as needed when clicked in the list and pass back the value to the parent component however now I am faced with another issue. I cannot for the life of me figure out how to set the default selected option. I have been at this for nearly 4 hours now and could use some guidance / help if anyone is available and has the time. Here's the current code: https://pb.farmflight.us/?6fbabf05594f72f9#C1EobuqQHPmx5WFJVN4sXw9Y2QauCZxjatvfvrKXgfuC - If you would like a video of what I am referring to, please let me know. Thank you!
